virttest/qemu_devices/qcontainer.py

```diff
@@ -1720,6 +1720,150 @@ def images_define_by_params(self, name, image_params, media=None,
+                else:
+                    pci_bus = {'aobject': 'pci.0'}
+                bus = self.get_first_free_bus(
+                    {'type': 'SERIAL', 'atype': serial_type}, [None, nr])
+                if bus is None:
+                    _hba = serial_type.replace('-', '_') + '%s'
+                    bus = self.idx_of_next_named_bus(_hba)
+                    bus = self.list_missing_named_buses(_hba, 'SERIAL', bus + 1)[-1]
+                    logging.debug("list missing named bus: %s", bus)
+                    devices.append(
+                        qdevices.QDevice(serial_type,
+                                         {"id": bus},
+                                         bus,
+                                         pci_bus,
+                                         qdevices.QSerialBus(bus, serial_type, bus)))
+                else:
+                    bus = bus.busid
+            devices.append(
+                qdevices.QDevice(port_type,
+                                 {"id": serial_id},
+                                 parent_bus={'busid': bus}))
+            devices[-1].set_param('name', serial_name)
+        else:  # none virtio type, generate serial device directly
+            devices.append(qdevices.QDevice(serial_type, {"id": serial_id}))
+            devices[-1].set_param("reg", reg)
+        devices[-1].set_param('chardev', chardev_id)
+
+        return devices
+
+    def serials_define_by_params(self, serial_id, serial_params,
+                                 file_name=None):
+        """
+        Wrapper for creating serial device from serial params.
+
+        :param serial_id: id of serial object
+        :param serial_params: serial params
+        :param file_name: the file path of the serial device (optional)
+        :return: the device list that construct the serial device
+        """
+        serial_type = serial_params['serial_type']
+        machine = serial_params.get('machine_type')
+
+        # Arm lists "isa-serial" as supported but can't use it,
+        # fallback to "-serial"
+        legacy_cmd = " -serial unix:'%s',server,nowait" % file_name
+        legacy_dev = qdevices.QStringDevice('SER-%s' % serial_id,
+                                            cmdline=legacy_cmd)
+        arm_serial = (serial_type == 'isa-serial'
+                      and 'arm' in serial_params.get("machine_type", ""))
+        if (arm_serial or not self.has_option("chardev")
+            or not self.has_device(serial_type)):
+            return legacy_dev
+
+        virtio_port_type = serial_params.get('virtio_port_type')
+        if virtio_port_type:
+            if "-mmio" in machine:
+                controller_suffix = 'device'
+            elif machine.startswith("s390"):
+                controller_suffix = 'ccw'
+            else:
+                controller_suffix = 'pci'
+            serial_type = '%s-%s' % (serial_type, controller_suffix)
+            virtio_port_type = "virt%s" % virtio_port_type
```

Comment:
how about use `virtserialport`, `virtconsole` as `virtio_port_type` ?  It is more easy to understand as it is same with qemu help output.  And then no need to add prefix `virt` here.

Reply:
@yanan-fu `virtio_port_type` is previously defined as `serialport` and `console`, just want to adopt previous usage.

Reply:
@vivianQizhu  I understand your point. 
Previously,  `console`, `virtio_console`, `serialport`, `virtio_serial` all can use,  since framwork accept all of them, and convert them to acceptable port type of qemu command line.
```
port_params.get('virtio_port_type') in ("console", "virtio_console"):
```
```
if 'console' in params.get('virtio_port_type'):
                port_type = 'virtconsole'
            else:
                port_type = 'virtserialport'
```

For me, it is confused.   If `console` , `virtio_console` can only be `virtconsole`,  and `serialport`, `virtio_serial` can only be `virtserialport`,  use the final port type is more clearly.
While in your another patch https://github.com/autotest/tp-qemu/pull/1434,  you have correct `virtio_serial`  to `serialport`, i think we can update them together.

Reply:
@yanan-fu Yes, it makes sense, I will update them.

Reply:
@yanan-fu There is variants named `virtserialport` and `virtconsole` in test case, I am afraid there is name conflict if modified them, I will need rename the variants and reorganize those `only` `no` keyword for them. I think leave them there is better, not really necessary to reorganize all params, which makes no improvement, just for save this postfix adding. Thanks.

Reply:
@vivianQizhu  oh~  i see those variants,  if so, it is ok to use `console` and `serialport`.